Allow configurable etcd options

jamiehannaford opened the pull request:
**What this PR does / why we need it**:

Allows users to set the `--listen-client-urls` and `--advertise-client-urls` flags on etcd binaries for clusters set up with kubeadm.

**Which issue this PR fixes**:

As far as I can tell right now, other nodes in a cluster set up with kubeadm cannot communicate with the etcd static pod running on the master. This is needed in order to set up calico/canal SDN which needs access to a publicly addressable IPv4 _before_ the overlay network and inter-cluster subnet is created.

Addresses https://github.com/kubernetes/features/issues/138 and https://github.com/kubernetes/features/issues/11.

**Release note**:
```release-note
Users can now specify listen and advertise URLs for etcd in a kubeadm cluster 
```

Commit:
Allow configurable etcd options

jamiehannaford commented:
The overall solution to this problem I think is to create a NodePort etcd service (either manually or automatically with kubeadm) and have the binary bind to `0.0.0.0`.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/42246)
<!-- Reviewable:end -->

pires commented:
@kubernetes/sig-cluster-lifecycle-pr-reviews @luxas @lukemarsden

jamiehannaford commented:
@pires @kubernetes/sig-cluster-lifecycle-pr-reviews @luxas @lukemarsden I know kubeadm is in code freeze due to moving over codebases - should I wait until that's happened or is there anything I  can incorporate now?

luxas commented:
@jamiehannaford Kubernetes as a project is in code freeze right now and no non-bugfix PRs can be merged at this point in the cycle. Hence the pending state...

kubeadm will eventually move to its own repo, but that will take some time yet before our machinery is able to do that.

jamiehannaford commented:
@luxas I know but some of my other PRs have been reviewed but not merged. Maybe it will save time when the code freeze is lifted. I'd just like a general idea of whether this feature is approved/needed and if any of the implementation logic is off :)

pires commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -108,6 +108,14 @@ func NewCmdInit(out io.Writer) *cobra.Command {
+		&cfg.Etcd.AdvertiseURL, "etcd-advertise-url", cfg.Etcd.AdvertiseURL,
+		"Specify a advertise-client-urls value to the etcd container",
+	)
 
```
Use configuration, not flags.

pires reviewed (changes requested):
Overall this LGTM. I just don't like the flags.

jamiehannaford commented:
@pires Does this look okay now? I removed the flags

pires commented on cmd/kubeadm/app/master/manifests.go:
```diff
@@ -367,6 +362,24 @@ func getAPIServerCommand(cfg *kubeadmapi.MasterConfiguration, selfHosted bool) [
+	if cfg.Etcd.AdvertiseURL != "" {
+		advertiseURL += fmt.Sprintf(",%s", cfg.Etcd.AdvertiseURL)
+	}
+
```
`listenURL` and `advertiseURL` should point to a same constant value.

jamiehannaford replied:
I wasn't 100% sure what you meant, but I added a new default variable which both URLs point to. That DRYed it up a bit. Is that what you meant?

pires commented:
/lgtm

pires reviewed (approved):
/lgtm

pires commented:
Ah, yes, the API changes break stuff. Please, check the test failure details and fix it.

luxas commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -78,10 +78,12 @@ type Networking struct {
+	CAFile       string
+	CertFile     string
+	KeyFile      string
+	ListenURL    string
```
I think we really should go with a generic Args here instead of two hard-coded string fields.
Imagine I want to set the token authz mechanism. Eventually we would duplicate all etcd options, instead we should go with a more dynamic route with `map[string]string` as the other control plane components

luxas commented on cmd/kubeadm/app/master/manifests.go:
```diff
@@ -367,6 +362,27 @@ func getAPIServerCommand(cfg *kubeadmapi.MasterConfiguration, selfHosted bool) [
+func getEtcdCommand(cfg *kubeadmapi.MasterConfiguration) []string {
+	defaultURL := "http://127.0.0.1:2379"
+
+	listenURL := defaultURL
```
Please use the same logic as the other control plane components here with default and overridable arguments.

luxas reviewed (commented):
There's nothing wrong with this PR per se, but I really think that if we do this, it should be generic from the get go.

Please align the logic here with what we have for all the other control plane components

jamiehannaford commented:
@k8s-bot bazel test this

jamiehannaford commented:
@luxas Thanks for the feedback, I've made things a bit more generic so folks can easily extend. Would you mind kicking off the tests again?

pires commented:
@k8s-bot bazel test this

luxas commented on cmd/kubeadm/app/master/manifests.go:
```diff
@@ -367,6 +362,20 @@ func getAPIServerCommand(cfg *kubeadmapi.MasterConfiguration, selfHosted bool) [
+		"advertise-client-urls": "http://127.0.0.1:2379",
+	}
+
+	command = append(command, "etcd", "--data-dir=/var/lib/etcd")
```
I would be ok with treating the data dir special; after all that's the heart of our cluster.
So for me you could add `DataDir string` to the struct and use it everywhere (do a search for /var/lib/etcd and use that val there instead).

You can make `/var/lib/etcd` a constant and use it as the default across the board.

luxas replied:
In particular I want to get rid of the etcd dir env param and make the volume mount actually respect what `DataDir` is set to

luxas commented on cmd/kubeadm/app/master/manifests_test.go:
```diff
@@ -624,6 +624,41 @@ func TestGetControllerManagerCommand(t *testing.T) {
+			},
+		},
+		{
+			cfg: &kubeadmapi.MasterConfiguration{Etcd: kubeadmapi.Etcd{ListenURL: "http://10.0.1.10:2379", AdvertiseURL: "http://10.0.1.10:2379"}},
```
Compile errors

jamiehannaford commented:
@luxas Agree about the dataDir stuff. I've added a new config key and removed the hard-coded references. Can you comment with `ok to test` so I can re-run tests to get them green?

dims commented:
@k8s-bot ok to test

jamiehannaford commented:
@k8s-bot non-cri e2e test this

luxas commented:
@k8s-bot non-cri e2e test this

jamiehannaford commented:
Looks like this is all green now

pires commented:
/lgtm

luxas commented on cmd/kubeadm/app/apis/kubeadm/env.go:
```diff
@@ -31,7 +33,7 @@ func SetEnvParams() *EnvParams {
 	envParams := map[string]string{
 		"kubernetes_dir":  "/etc/kubernetes",
-		"host_etcd_path":  "/var/lib/etcd",
+		"host_etcd_path":  v1alpha1.DefaultEtcdDataDir,
```
Sorry for being unclear earlier, I don't think this is used anymore (if it is, switch to the new method)
It could/should be removed

jamiehannaford replied:
No problem, I appreciate the feedback 👍  does this commit address what you mean? https://github.com/kubernetes/kubernetes/pull/42246/commits/4b065f8a27a08d70b0e39b2ef99af44f35689859

luxas reviewed (commented):
LGTM, only one nit.
Thanks for doing this!

jamiehannaford commented:
@k8s-bot non-cri e2e test this

pires commented:
/lgtm

luxas commented:
@jamiehannaford LGTM, but please squash (unless you really think the commit sequence makes a lot of sense)

jamiehannaford commented:
@luxas Sure - I squashed those commits into 1

jamiehannaford commented:
Any update on this?

luxas reviewed (approved):
Hi, thanks for bearing with us. I've been away last week, digging myself up now from all the emails.

Thanks for the ping!

LGTM

luxas commented:
/lgtm
/approve

jamiehannaford commented:
@k8s-bot gce etcd3 e2e test this
